compiler/rustc_hir_typeck/src/lint.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_hir::def_id::DefId;
+use rustc_hir::{HirId, HirIdSet};
+use rustc_middle::hir::place::{PlaceBase, PlaceWithHirId};
+use rustc_middle::mir::FakeReadCause;
+use rustc_middle::query::Providers;
+use rustc_middle::ty::{self, TyCtxt};
+use tracing::instrument;
+
+use crate::expr_use_visitor::{Delegate, ExprUseVisitor};
+
+#[derive(Default)]
+struct ExtractConsumingNodeDelegate {
+    nodes: HirIdSet,
+}
+
+impl<'tcx> Delegate<'tcx> for ExtractConsumingNodeDelegate {
+    #[instrument(level = "debug", skip(self))]
+    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, _: HirId) {
+        match place_with_id.place.base {
+            PlaceBase::Rvalue => {
+                self.nodes.insert(place_with_id.hir_id);
+            }
+            PlaceBase::Local(id) => {
+                self.nodes.insert(id);
+            }
+            PlaceBase::Upvar(upvar) => {
+                self.nodes.insert(upvar.var_path.hir_id);
+            }
+            PlaceBase::StaticItem => {}
+        }
+    }
+
+    fn borrow(
+        &mut self,
+        _place_with_id: &PlaceWithHirId<'tcx>,
+        _diag_expr_id: HirId,
+        _bk: ty::BorrowKind,
+    ) {
+        // do nothing
+    }
+
+    fn mutate(&mut self, _assignee_place: &PlaceWithHirId<'tcx>, _diag_expr_id: HirId) {
+        // do nothing
+    }
+
+    fn fake_read(
+        &mut self,
+        _place_with_id: &PlaceWithHirId<'tcx>,
+        _cause: FakeReadCause,
+        _diag_expr_id: HirId,
+    ) {
+        // do nothing
+    }
+}
+
+fn extract_tail_expr_consuming_nodes<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx HirIdSet {
```

Comment:
Make this into a `hook`, not a `query`; I don't believe it needs any sort of caching, since that has overhead. This should also explain very clearly what it does.